Merge documentation conversion branch to master branch

baskaufs opened the pull request:
As of 2018-07-18, the four major documents of Audubon Core (introduction, structure, term list, and guide) have been converted to Markdown.  In the case of the term list, the Markdown document is generated by prepending a header document and appending a footer document to the actual term list generated from CSV data in the https://github.com/tdwg/rs.tdwg.org repo.  

Because the conversion to Markdown has not substantively changed the contents of the documents, I would assert that new versions have not been generated, although this should probably be discussed.

Remaining on 2018-07-18 to be completed is editing the introductions of all four of these documents to indicate what portions of each are normative.

baskaufs commented:
Based on the discussion at the 2018-07-18 hangout, it seems clear that the introductory document 

https://github.com/tdwg/ac/blob/documentation-conversion/doc/introduction.md

should be demarcated as having no normative content.  It is possible that at some future point this document might be rendered unnecessary if the AC Standard landing page supplants it.  However, that's a problem that would require invoking the change process and creating a new version of the standard, so for now I think we should let it be and simply bring it into conformance with the Standards Documentation Specification.

If there is no objection to this by the end of the day Friday, 2018-07-27, I'll make the notation and close this.

baskaufs commented:
Based on the discussion at the 2018-07-18 hangout, it seems clear that the "guide" document, 

https://github.com/tdwg/ac/blob/documentation-conversion/doc/guide.md

whose current introduction describes (and has always described) it as non-normative, should be clearly demarcated as having no normative content.  If there are no objections by the end of the day Friday, 2018-07-27, I will delete the "note" paragraph in the introduction and replace it with a clear statement at the beginning of the section stating that it is entirely non-normative, and include updated links to the Markdown serializations of the two documents that DO contain normative content.

baskaufs commented:
Based on the discussion at the 2018-07-18 hangout, it seems clear that most parts of the "structure" document

https://github.com/tdwg/ac/blob/documentation-conversion/doc/structure.md

should be considered normative, since its introduction states that it is normative documentation for the standard.  However, for ease of maintenance, I believe that the statement required by the Standards Documentation Spec about what parts should be normative should exclude examples.  I propose clearly labeling the examples as "Example 1", etc. and then noting in the statement that Sections 2 through 4 are normative, excluding examples.

This document uses the words "must", "may", "recommended", etc. without capitalization, or any indication that the use of those words implies the meanings of RFC 2119.  The Standards Documentation Spec does not require using RFC 2119, it only says that if the meanings of RFC 2119 are intended that must be stated in the document.  I would like opinions as to whether there are any places within this document where capitalization and application of RFC 2119 would be appropriate.  My default assumption is that there are no such places, since RFC 2119 wasn't invoked in the original document.  However, since this will be the last pass on the document before its conformity with the SDS will be reviewed, I'd like confirmation that we don't think that RFC 2119 should apply anywhere.  If we DO think that RFC 2119 should be invoked, then I'm a bit unsure as to whether the change process needs to be invoked.  My guess is that it should.  

If there are no objections by the end of the day Friday, 2019-07-27, I'll make the suggested edits and consider this resolved.  If we think that RFC 2119 needs to be invoked, then I'll open that as an issue in the tracker and we will probably have to work on it well beyond Friday.

nielsklazenga commented:
I've gone through the introduction, structure and guide documents so far:

* Pandoc escapes less than (<) and underscores, which shouldn't be escaped. Made two commits to fix it, but found I don't have write access to the tdwg/ac repo. Will email Peter and Stan and ask them to remedy this, so hopefully they'll arrive soon. Otherwise I will do a pull request.
* Several of the links in doc/structure.md and doc/guide.md are now broken. Should I put in the new links? I think it is probably best, for now, to do this as annotations, rather than replace the old link. 
* doc/introduction.md has a link to the Audubon Core Implementations page on the old wiki. This page hasn't been converted to MarkDown yet. Should I do that, or was it left out on purpose?
* I can't find the Audubon Core Guide document on the old wiki, so I can't check a few things that look a bit suspect. Can someone get the link for me?

Moving on to the Term List...

baskaufs commented:
The “guide” document is here:
https://terms.tdwg.org/wiki/Audubon_Core_Non_Normative_Document
It’s only available as a Word document.  

The position that I’m taking (based section 3.2.2 of the Vocab Management Spec) is that our maintenance group can fix broken links at any time without invoking any kind of process.  So if there are corrected links, I’d just fix them.  I hadn’t actually worked on that yet, partly because some of the links may be to documents that will get moved around as we organize the documents within the AC github repo.  I was figuring that when we have merged the document conversion branch and have everything where we think it belongs, we would make a final pass at making sure all of the links are correct.  

As far as the “implementations” page is concerned, I didn’t convert it yet because I hadn’t finished thinking through whether it should continue to exist in its current form.  In the end, what we need to have is a landing page for the standard that states clearly what documents are officially part of the standard.  We should also have some page (maybe the landing page or maybe something else) that will lead users to other helpful information that is not part of the standard itself.  I think that with the adoption of the Standards Documentation Specification, the RDF document is obsolete except as a historical item.  I’m not sure about the Audubon Core Development document.  I haven’t really looked at it, but I think the SDS clarifies the role that SKOS should play in TDWG vocabularies.  So I suspect that it may also be obsolete.  That only leaves the IPT extension, to which there should be a link.  But I doubt we need a special implementations page for that one link.  I guess I was thinking of just ignoring the AC Implementations page until the issue of what’s on the landing page gets settled.

One thing that’s unsettled is what web page will actually be the landing page.  I was assuming that it would be the README.md page for the root directory of the Github repo.  That’s what it is currently (i.e. the permanent vocabulary IRI dereferences to it).  But now that there is a new TDWG website, it might actually live there.  That’s another reason for maybe putting off working on the standard landing page until last.  I think this is high on Stan’s agenda since they want to get the new website launched, so hopefully there will be some clarity on that soon.

nielsklazenga commented:
Okay, I'll just fix the links. They are all external links, so I might as well do it now.

Regarding the Implementations document, that was exactly what I was thinking, which is why I asked. I'll remove the reference from the Introduction document.

baskaufs commented:
Sounds good, @nielsklazenga .  If you get write access, just change the document-conversion branch.  If not, just make a pull request.  But really, all of the core members of the maintenance group should have write access, so let's push for that.

nielsklazenga commented:
I've got write access now.

baskaufs commented:
Here is a brief summary of my edits to the "introduction", "structure", and "guide" documents:

- In each document, I added a subsection of the introduction called "Status of the content of this document".  In that subsection, I indicated which parts were normative and non-normative.  ([Section 3.2.1 of the SDS](https://github.com/tdwg/vocab/blob/master/sds/documentation-specification.md#32-descriptive-documents)).

- In each document, I edited the introduction as necessary to explain the purpose of the document ([Section 3.2.3.3 of the SDS](https://github.com/tdwg/vocab/blob/master/sds/documentation-specification.md#32-descriptive-documents)).  I also removed the hyperlinks to the other three documents that I had temporarily stashed at the top of the document and placed them in the introduction so that if a user discovered that they were in the wrong document, they could easily go to the right one.

- In the [guide document](https://github.com/tdwg/ac/blob/documentation-conversion/doc/guide.md) I removed the historical information from the introduction and moved it to Appendix II, which already contained other historical information.  It was a distraction to begin the introduction with historical details rather than an explanation of the purpose and scope of the document.

- In the structure document, I moved around some content in [Section 3](https://github.com/tdwg/ac/blob/documentation-conversion/doc/structure.md#3-multiplicitycardinality) so that the examples could be separated from the text that described them.  Originally, the entire document was described as normative.  In order to separate the normative content from the content that I was declaring as non-normative (the examples) they had to be in separate subsections as required by [Section 3.2.1](https://github.com/tdwg/vocab/blob/master/sds/documentation-specification.md#32-descriptive-documents) of the SDS.  

- I also corrected some incorrect hyperlinks in [Section 4](https://github.com/tdwg/ac/blob/documentation-conversion/doc/introduction.md#4-references) of the Introduction document and probably some other places when I found them.

- I did not perform a detailed proofreading review of the [guide document](https://github.com/tdwg/ac/blob/documentation-conversion/doc/guide.md).  That's because I don't think a general revision of the AC documents is within the scope of bringing the standard into conformance with the SDS.  However, a vocabulary maintenance group can at any time fix errors like bad hyperlinks, so I think at a minimum, we still should go through the guide document to look for outdated references to documents that no longer exist or whose URLs have changed.  For example I noticed references to the deprecated RDF documents.  I'll open an issue about this so we don't forget to fix it.

baskaufs commented:
Thanks for going through this, Niels.  I think the development page can go away.  I think there was an earlier vision of how AC would be maintained via the wiki, but that has sort of fallen by the wayside.  

As far as the links are concerned, as long as they work, the simpler the better.  I think that what happened was that I did a bunch of testing to try to figure out how to get links within the document to work and the only way I could get them to work was with HTML.  So if there is a simpler way to do it using only Markdown, we should go for it.  That would probably require changing the script, but those changes would be trivial.

As far as links 3 and 6 are concerned, if one can find new links by Googling, we can change them.  If not, then I'm not sure what to do since they are recommended sets of controlled vocabularies.

It is possible that those footnotes will just go away.  In the wiki version of the page, they were linked from the Iptc4xmpExt:CVterm Notes section.   But if we want the notes to be independent of the structure of the human-readable document, then we probably should just include the hyperlinks directly in the notes instead of putting them at the end.  In that case, the footer would only include the licensing info.

nielsklazenga commented:
I made those changes. The raw form is now much easier to read. I've removed the bookmarks form the headings and use the ones GitHub (Jekyll) creates automatically. It would be good to remove them from the Python script as well. The bookmarks for the terms should stay, as GitHub will remove the colon and not replace it with something else; it would be nice if we could use hyphens rather than underscores though.

I found alternatives for the broken links in the References, but noticed that all the citations are under [Iptc4xmp:CVterm](https://github.com/tdwg/ac/blob/documentation-conversion/doc/termlist.md#Iptc4xmpExt_CVterm), which comes from the CSV file, which doesn't have the links. It would be good to have the script put them in. I will find out how footnotes can be done in Markdown (the references were MediaWiki footnotes).

I will be happy to play around with the script, but that would have to wait till the weekend, as I am not too familiar with Python, so it will take me much longer than it should.

Will clean up the HTML links in the other documents.

nielsklazenga commented:
Replicating the Mediawiki endnotes in Markdown is a pain and can't be done without using HTML. Look what I've done in the [Guide](https://github.com/tdwg/ac/blob/documentation-conversion/doc/guide.md) document. If we can do without the back-link from the endnote to the citation, it will be much easier, as then the links to the endnotes are just internal links.

baskaufs commented:
Thanks for making these edits, @nielsklazenga .  As I thought about the issue of the hyperlinks some more, I think the actual reason why I made them using HTML rather than using the automatic linking was in order to have the element IDs/fragment identifiers remain the same as they were in the mediawiki document.  I was thinking that if at some point in the future there were a redirect from the mediawiki URL to the Markdown URL, bookmarks of URLs containing the fragment identifiers would not be broken.  I'm not sure whether that would be true, and at this point I'm thinking it may not be important enough to override the much cleaner markdown you've created.

I don't have time today to edit the script to follow the pattern you've set up, but I'll try to do it in the next several days.  So you don't need to do it.  

As far as the endnotes are concerned, I can just make them go away by putting the URLs into the dcterms:description field for the Iptc4xmp:CVterm term in the table that generates the human-readable term metadata.  Again, I don't have time to do it today, but it's a trivial edit and I should be able to do it soon.  So I think we can just forget about the endnotes in the footer - I will make them go away.

nielsklazenga commented:
I'll wait until you're done. Good idea to put those URLs in the CSV: I had a quick look at the pages yesterday and they are quite useful.

Could I suggest that we rename the 'doc' folder to 'docs'? GitHub can use the 'docs' folder as the root for the https://tdwg.github.io/ac pages, which means that we don't have to put HTML files in the gh-pages branch, as is done in the 'dwc' repo.

baskaufs commented:
I made the change to the rs.tdwg.org CSV file so that the CVterm description includes the hyperlinks.  However, I realized that endnote 9 comes from some text that doesn't even get inserted by the script.  So there is still some work to be done on that.  But the generated termlist document has the corrected links now.

baskaufs commented:
Footnotes have been rendered unnecessary in the termlist document and the term section comments are now where they should be in the termlist document (Geographic Coverage and Service Access Point sections).

baskaufs commented:
@nielsklazenga I noticed that some of the hyperlinks that were changed to native Markdown don't work.  I think in many cases, that's because of missing "#" characters.  I haven't gone through and tried all of the links, but we should do that and look at the Markdown to see what needs to be fixed if the links don't work.  I think all of the ones in the termlist.md document are OK, but not necessarily in the other 3 documents.

nielsklazenga commented:
@baskaufs Those will be typos of mine, as they really need to have the hash. Will check.

nielsklazenga commented:
That should be fixed now @baskaufs. Also converted some really messy tables in doc/guide.md into HTML, which should make them easier to read and maintain (also lost the empty header row).

baskaufs commented:
@nielsklazenga  I've renamed the "doc" folder to "docs" to make Jekyll theming/management possible, as you suggested.  I'm not up on that (yet) but it seems like a good thing for the future.

nielsklazenga commented:
@baskaufs: I am. Have never done it on a docs folder yet though. Will try on a fork first.

nielsklazenga commented:
@baskaufs Pages are now up at https://nielsklazenga.github.io/ac/. Note that this is just the out-of-the-box Jekyll blog design. I could easily put in the new TDWG website design, but we should talk with @peterdesmet first and see how he wants to do it. Jekyll automatically changed the links between pages to the new links, which I thought we might have to do manually. On the other hand, in the indexes, I had to change the double pipes to single ones, as they were interpreted as tables.

baskaufs commented:
@nielsklazenga That looks really great!  I think we should just make the switch.  I don't think that it's important that the styling is done now - that can always be improved later.  I would mostly like to get the content into place before the meeting if we can, which I think is possible.

Two questions: 
1. I suppose we can control what shows up at the bottom of the page, right?  I would prefer for us to just explicitly control what is in the footer rather than having Jekyll automatically put in stuff like the Twitter link.

2. We should think carefully about the "permalink" URL endings.  I'm assuming that you used index.html so that the Introduction page will default there if there if the last part of the URL is left off.  Ultimately, the behavior should be that a redirect from the "Latest version:" URL will go to the page.  So the end of the URL generated by Jekyll should match with some pattern that can be computed by server script from the end of the Latest version URL.  Probably that means that introduction should end in "/introduction"

nielsklazenga commented:
Hi @baskaufs,
1. I just put the TDWG website layout in (not everything is working yet). Yes, we can control all that.
2. Yes, I needed a page at the base URL, so I set the permalink for the introduction to '/index.html'. Darwin Core does the same. That's why I also set the permalink for the term list to '/terms'. But yes, we could put in anything in the permalink.

I think it is time someone pushed the 'Merge pull request' button.

baskaufs commented:
I think I'd like to fix the scopeNote thing before we do the merge.  But after that I think things will be good enough for the edits to go the the master branch.  The danger of merging too soon is for people to think that what we are displaying is the authoritative version.  So it really needs to be close to right before people start seeing it there.

baskaufs commented:
See latest comments on Issue #109 about the addition of scope notes

nielsklazenga commented:
Funny, that was like the one bit of the script I didn't touch. I'll check it out.

nielsklazenga commented:
> I had a problem seeing how the Iptc4xmpExt:CVterm links rendered after being processed by the createLinks function that you added to the script. So I commented it out for the time being so that I could make sure everything was there in the raw Markdown. Once we do the merge and get everything into the master branch, we can see how Jekyll renders the documents in the docs directory. I don't understand how the <i class="fa fa-external-link"></i> tags work, but we can play with turning the createLinks function back on again (subject to what I said above about really wanting the Notes metadata to match with what will get served as machine-readable metadata).

'<i class="fa fa-external-link"></i>' is the font-awesome icon for external link, similar to the Mediawiki one. How about we put the links back, but use the URL as the label (as well as the href)? I can then write a Javascript function to replace the URL label with the font-awesome icon and have the URL in the title attribute.

> Oh, there was one more thing. I see that you changed the end of the URL of the termlist document to "/terms". That is probably OK and maybe better than "/termlist". However, I was hoping that the document name we end up with at the end of the URL would match with the permanent URL for the document. Currently that's http://rs.tdwg.org/ac/doc/termlist/ . Also, "docs" doesn't match with "doc", but that isn't so critical. But people might somehow get confused, so in the end, we should try to make the permanent URLs correspond as much as possible to the ends of the URLs that we will have at github.io .

I wasn't really aware of the existing http://rs.tdwg.ac/doc URLs. I'll change the permalinks to reflect the filenames. The 'docs' bit can easily be taken care of in the URL rewrite. GitHub needs it to be 'docs' to put the pages up. The 'docs' is not actually in the URL, so we could have a 'doc' folder within 'docs'. Might actually try that; also has the benefit that all the Jekyll stuff I need to add does not sit among our documents.

I am going to make use of the time difference between us and get ready before you do any more commits.

baskaufs commented:
@nielsklazenga Just do what you think works best and we can always change it back if it doesn't work.  The Javascript fix sounds good to try since the markdown itself would still have the links.  the doc/docs thing isn't that important.  I wasn't sure whether to use "doc" or "docs" and could just switch them all.  You can see the pattern I used here: https://github.com/tdwg/rs.tdwg.org/blob/master/docs/docs.csv .  But I could just change all of the "doc" to "docs" since none of this has been implemented yet.

baskaufs commented:
@nielsklazenga This really looks amazing!  Thanks for your work on all of the Jekyll styling.

Should we delete the documentation-conversion branch or leave it?

At this point I'm going to do some work on the README.md page at the root of the ac repo and try to figure out how to work on the AC landing page on the new website.
